### ISSUE TITLE
Make Analytics property lookup scalable for matching property

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#341]
## Bug Description

Currently, Analytics properties are queued up per Analytics account (necessary per API limitations) and compared with either an existing tracking ID or the website URL. This approach is not scalable if somebody has too many Analytics accounts, as it happens via individual requests (from Site Kit client to Site Kit server, and from Site Kit server to Google API).

In addition, this problem is related to some of the existing Analytics datapoints covering additional logic to what they claim to do by their name. For example,  `get-accounts` contrary to its name does not only return accounts, but also tries to detect a matching set of properties (although not covering the entire spectrum, e.g. when an existing tag exists, see #311) and return that too.

I suggest the following:
* Remove the logic for finding the matching account from `get-accounts`. Simply get the accounts, as the route name indicates, nothing more.
* In JavaScript, after getting the accounts do the following:
    * Build a batch request, calling the `get-properties` datapoint with every individual `accountId`. If more than 10 accounts, only include the first 10 in the batch request.
    * Iterate through the properties in all responses:
        * If an existing tag is present, find the matching property by comparing the tracking ID.
        * If no existing tag is present, find the matching property by comparing the website URL.
    * If a property was matched, we now can lock to that property (if existing tag) or preselect that property (if no existing tag).
    * If no property was matched, we continue with the next <= 10 accounts (if it was more than 10 originally), and so on. If we completed checking for all accounts and there is still no matching property, we proceed as usual: If an existing tag, it means the user doesn't have permission. If no existing tag, it means there's no property that is associated with the website URL.

This has several advantages:
* The approach is scalable, even if someone had 200 accounts, it wouldn't result in a timeout (today it likely would in multiple server-side areas).
* API routes are simplified and do what their names indicate.
* The `tag-permission` route is no longer needed.

Auxiliary requirements for making this change:
* In the `combinedGet()` function, the key must be generated in a more flexible way, supporting any data parameters, not just a few hard-coded ones (`permaLink`, `dateRange`).

**This is blocked by #83.**

cc @lucymtc 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

In Analytics:
* There should be an `accounts` datapoint that returns a list of accounts, nothing else.
* There should be a `properties` datapoint that returns a list of properties for a given `accountID`, nothing else.
* If there is an existing tag, ...
    * the plugin should query the `properties` datapoint with `accountID` set to the account ID extracted from the existing tag (i.e. `UA-{account.id}-x`)
    * if properties include the exact property ID from the tag, the user has access (and the property should be force-selected, i.e. the dropdown be disabled)
    * else they're blocked
* else ...
    * the plugin should query the `accounts` datapoint to get all available accounts
    * the plugin should fire requests to `properties` for each account found, and after every request check if a property matches the website URL
    * whenever a match is found, stop firing more requests and pre-select the corresponding account and property in the dropdown (but _do not_ force-select it, the user should be able to change it)
    * if there is no match found after all `properties` requests, don't preselect anything, leave "Select an account" and "Select a property..."
* generally, if a property is preselected (or force-selected/locked), a request for the corresponding `profiles` should be made, to populate the third dropdown

To limit the amount of `properties` requests issued to match the website URL, let's set a threshold (e.g. 5). If there are more accounts than that threshold, we're not going to try to find a match. In that case, in the `else ...` above we should only do the first step of querying `accounts` once.

## Implementation Brief

* 

## Changelog entry

*

**Comment by aaemnnosttv** [comment]
@felixarntz in the case of an existing tag, I think we can bypass the account crawling as the account number is part of the property ID as `UA-{account.id}-{profile.num}`. We should be able to just query the account specifically. 

As for URL matching, it would be great if there was a way to query web profiles by `websiteUrl` across all accounts your user has access to but it doesn't look like that's possible.

**Comment by felixarntz** [comment]
@aaemnnosttv Do you mean properties instead of profiles? The profiles (or views, synonymous) are not relevant here I think. Basically, the `property.id` is `UA-{account.id}-{property.num}` I think. But yeah, we should thus be able to query for that account, and then check for the property specifically in that account (if accessible).

>  it would be great if there was a way to query web profiles by `websiteUrl`

Yeah, that doesn't exist unfortunately.

**Comment by felixarntz** [comment]
@aaemnnosttv I updated the ACs to something more applicable - what do you think?

**Comment by ThierryA** [comment]
@felixarntz could you provide more info as why this was put in the "Revisit Later" column?

**Comment by felixarntz** [comment]
This issue has been in that column forever due to #83, which however is long done. It can be moved to AC column for finalizing those.

**Comment by felixarntz** [comment]
Some related changes are being made in #1316 - TLDR if we have a property ID, we shouldn't need to search for the account for it.

For finding a property by matching a URL, that would still require multiple API queries, one per account.

**Comment by felixarntz** [comment]
I'll close this one, since the effort to make it work given the impact (it's unlikely that any user has _that_ many Analytics accounts in their Google account) would not be worth it.
